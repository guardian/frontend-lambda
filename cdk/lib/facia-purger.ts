import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import { GuLambdaFunction } from '@guardian/cdk/lib/constructs/lambda';
import type { App } from 'aws-cdk-lib';
import { CfnParameter } from 'aws-cdk-lib';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { EmailSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';

export class FaciaPurger extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const fastlyServiceId = new CfnParameter(this, 'FastlyServiceId', {
			type: 'String',
			description: 'Id of service to purge',
		});
		const fastlyAPIKey = new CfnParameter(this, 'FastlyAPIKey', {
			type: 'String',
			description:
				'API key with purge writes for service with id FastlyServiceId',
			noEcho: true,
		});
		const emailRecipient = new CfnParameter(this, 'EmailRecipient', {
			type: 'String',
			description: 'The email address to send alarm notifications to',
		});

		const notificationTopic = new Topic(this, 'NotificationTopic');
		notificationTopic.addSubscription(
			new EmailSubscription(emailRecipient.valueAsString),
		);

		const lambda = new GuLambdaFunction(this, 'faciaPurgerLambda', {
			app: 'facia-purger',
			fileName: 'facia-purger.jar',
			runtime: Runtime.JAVA_11,
			handler: 'com.gu.purge.facia.Lambda',
			environment: {
				FastlyServiceId: fastlyServiceId.valueAsString,
				FastlyAPIKey: fastlyAPIKey.valueAsString,
				Stage: this.stage,
			},
			memorySize: 512,
			errorPercentageMonitoring: {
				toleratedErrorPercentage: 1,
				numberOfMinutesAboveThresholdBeforeAlarm: 5,
				snsTopicName: notificationTopic.topicName,
			},
		});

		this.overrideLogicalId(lambda, {
			logicalId: 'Lambda',
			reason: 'preserve existing triggers',
		});
	}
}
