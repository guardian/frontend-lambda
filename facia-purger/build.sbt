name := "facia-purger"

scalaVersion := "2.11.8"

organization := "com.gu"
description := "Lambda for purging Fastly cache based on s3 events"

libraryDependencies ++= Seq(
  "com.amazonaws" % "aws-lambda-java-core" % "1.1.0",
  "com.amazonaws" % "aws-lambda-java-events" % "1.3.0",
  "com.squareup.okhttp3" % "okhttp" % "3.2.0",
  "org.parboiled" %% "parboiled" % "2.1.3",
  "org.apache.logging.log4j" % "log4j-core" % "2.16.0",
  "com.amazonaws" % "aws-lambda-java-log4j2" % "1.4.0",
  "org.scalatest" %% "scalatest" % "2.2.2" % "test",
  "org.mockito" % "mockito-all" % "1.9.5" % "test"
  )
  
def env(key: String): Option[String] = Option(System.getenv(key))

lazy val root = (project in file(".")).enablePlugins(RiffRaffArtifact)

riffRaffPackageName := "facia-purger"
riffRaffPackageType := assembly.value
riffRaffUploadArtifactBucket := Option("riffraff-artifact")
riffRaffUploadManifestBucket := Option("riffraff-builds")
assemblyJarName := normalizedName.value + ".jar"
riffRaffBuildIdentifier := env("BUILD_NUMBER").getOrElse("DEV")
riffRaffManifestBranch := env("BUILD_VCS_BRANCH").getOrElse(git.gitCurrentBranch.value)
riffRaffManifestProjectName := s"dotcom:lambda:${normalizedName.value}"
riffRaffArtifactResources := Seq(
  baseDirectory.value / "cloudformation.yaml" -> "cloudformation/cloudformation.yaml",
  baseDirectory.value / "riff-raff.yaml" -> "riff-raff.yaml",
  assembly.value -> s"${normalizedName.value}/${normalizedName.value}.jar"
)

assemblyMergeStrategy in assembly := {
  case PathList("META-INF", xs @ _*) => MergeStrategy.discard
  case _                             => MergeStrategy.first
}
