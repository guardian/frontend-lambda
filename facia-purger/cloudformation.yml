AWSTemplateFormatVersion: 2010-09-09
Description: Facia purger lambda
Parameters:
  PackageName:
    Description: Lambda name
    Type: String
    Default: facia-purger
  Stage:
    Description: Stage name
    Type: String
    AllowedValues:
      - CODE
      - PROD
    Default: CODE
  FastlyServiceId:
    Description: Id of service to purge
    Type: String
  FastlyAPIKey:
    Description: API key with purge writes for service with id FastlyServiceId
    Type: String
    NoEcho: true
  DeployBucket:
    Description: Bucket where RiffRaff uploads artifacts on deploy
    Type: String
  EmailRecipient:
    Description: The email address to send alarm notifications to
    Type: String
  PrivateSubnets:
    Type: List<AWS::EC2::Subnet::Id>
    Description: Private subnets of vpc
  VPC:
    Type: AWS::EC2::VPC::Id
    Description: VPC to deploy into
Resources:
  ExecutionRole:
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument:
        Statement:
          - Effect: Allow
            Principal:
              Service:
                - lambda.amazonaws.com
            Action: sts:AssumeRole
      Path: /
      Policies:
        - PolicyName: logs
          PolicyDocument:
            Statement:
              Effect: Allow
              Action:
                - logs:CreateLogGroup
                - logs:CreateLogStream
                - logs:PutLogEvents
              Resource: arn:aws:logs:*:*:*
        - PolicyName: s3
          PolicyDocument:
            Statement:
              Effect: Allow
              Action:
                - s3:GetObject
              Resource: arn:aws:s3:::aws-frontend-store/*
        - PolicyName: network-interfaces
          PolicyDocument:
            Statement:
              Effect: Allow
              Action:
                - ec2:DescribeNetworkInterfaces
                - ec2:CreateNetworkInterface
                - ec2:DeleteNetworkInterface
              Resource: "*"
  Lambda:
    Type: AWS::Lambda::Function
    Properties:
      Code:
        S3Bucket: !Ref DeployBucket
        S3Key: !Sub "frontend/${Stage}/${PackageName}/${PackageName}.jar"
      Description: Purge fastly after a facia change event from an S3 bucket
      Handler: com.gu.purge.facia.Lambda
      MemorySize: 512
      Role: !GetAtt ExecutionRole.Arn
      Runtime: java8
      Timeout: 30
      Environment:
        Variables:
          Stage: !Ref Stage
          FastlyAPIKey: !Ref FastlyAPIKey
          FastlyServiceId: !Ref FastlyServiceId
      VpcConfig:
        SubnetIds: !Ref PrivateSubnets
        SecurityGroupIds:
          - !Ref LambdaSecurityGroup
  NotificationTopic:
    Type: AWS::SNS::Topic
    Properties:
      Subscription:
        - Protocol: email
          Endpoint:
            Ref: EmailRecipient
  InvocationAlarm:
    Type: AWS::CloudWatch::Alarm
    Properties:
      InsufficientDataActions:
        - Ref: NotificationTopic
      AlarmDescription: Notify if there are less than 5 invocations over last 5 minutes or there is insufficient data (i.e. no invocations)
      ComparisonOperator: LessThanOrEqualToThreshold
      Dimensions:
        - Name: FunctionName
          Value:
            Ref: Lambda
      EvaluationPeriods: 1
      MetricName: Invocations
      Namespace: AWS/Lambda
      Period: 3600
      Statistic: Sum
      Threshold: 0
      Unit: Count
  ErrorAlarm:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmActions:
        - Ref: NotificationTopic
      AlarmDescription: Notify if there are more than 20 errors over last 5 minutes
      ComparisonOperator: GreaterThanOrEqualToThreshold
      Dimensions:
        - Name: FunctionName
          Value:
            Ref: Lambda
      EvaluationPeriods: 1
      MetricName: Errors
      Namespace: AWS/Lambda
      Period: 300
      Statistic: Sum
      Threshold: 20
      Unit: Count
  LambdaSecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: Security group for github lambda - allow access to fastly
      VpcId: !Ref VPC
      SecurityGroupEgress:
        CidrIp: 0.0.0.0/0
        IpProtocol: tcp
        FromPort: 443
        ToPort: 443
