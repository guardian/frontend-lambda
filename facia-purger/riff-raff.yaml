stacks:
  - frontend
regions:
  - eu-west-1
deployments:
  cloudformation:
    type: cloud-formation
    app: facia-purger-lambda
    parameters:
      cloudFormationStackByTags: false
      prependStackToCloudFormationStackName: false
      cloudFormationStackName: facia-purger-lambda
      templateStagePaths:
        PROD: cloudformation-PROD.yaml
        CODE: cloudformation-CODE.yaml
  facia-purger:
    type: aws-lambda
    parameters:
      bucket: aws-frontend-artifacts
      functions:
        PROD:
          name: facia-purger-lambda-PROD-Lambda-190WIIV8979JL
          filename: facia-purger.jar
        CODE:
          name: facia-purger-lambda-CODE-Lambda-1AI0OAR4BOJDC
          filename: facia-purger.jar
    dependencies:
      - cloudformation
