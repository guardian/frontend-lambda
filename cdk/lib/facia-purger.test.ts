import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { FaciaPurger } from './facia-purger';

describe('The FaciaPurger stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new FaciaPurger(app, 'FaciaPurger', {
			stack: 'frontend',
			stage: 'TEST',
		});
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
