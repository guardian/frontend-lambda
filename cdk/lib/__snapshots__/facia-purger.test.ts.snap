// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FaciaPurger stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "EmailRecipient": Object {
      "Description": "The email address to send alarm notifications to",
      "Type": "String",
    },
    "FastlyAPIKey": Object {
      "Description": "API key with purge writes for service with id FastlyServiceId",
      "NoEcho": true,
      "Type": "String",
    },
    "FastlyServiceId": Object {
      "Description": "Id of service to purge",
      "Type": "String",
    },
  },
  "Resources": Object {
    "Lambda": Object {
      "DependsOn": Array [
        "faciaPurgerLambdaServiceRoleDefaultPolicyD659707F",
        "faciaPurgerLambdaServiceRoleD80DA945",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "frontend/TEST/facia-purger/facia-purger.jar",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "facia-purger",
            "FastlyAPIKey": Object {
              "Ref": "FastlyAPIKey",
            },
            "FastlyServiceId": Object {
              "Ref": "FastlyServiceId",
            },
            "STACK": "frontend",
            "STAGE": "TEST",
            "Stage": "TEST",
          },
        },
        "Handler": "com.gu.purge.facia.Lambda",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "faciaPurgerLambdaServiceRoleD80DA945",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "facia-purger",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/frontend-lambda",
          },
          Object {
            "Key": "Stack",
            "Value": "frontend",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "NotificationTopicEB7A0DF1": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/frontend-lambda",
          },
          Object {
            "Key": "Stack",
            "Value": "frontend",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::SNS::Topic",
    },
    "NotificationTopicTokenSubscription180FA58AD": Object {
      "Properties": Object {
        "Endpoint": Object {
          "Ref": "EmailRecipient",
        },
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "NotificationTopicEB7A0DF1",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "faciaPurgerLambdaErrorPercentageAlarmForLambdaB0D3D9E0": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":",
                Object {
                  "Fn::GetAtt": Array [
                    "NotificationTopicEB7A0DF1",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "Lambda",
              },
              " exceeded 1% error rate",
            ],
          ],
        },
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "High error % from ",
              Object {
                "Ref": "Lambda",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 5,
        "Metrics": Array [
          Object {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "Error % of ",
                  Object {
                    "Ref": "Lambda",
                  },
                ],
              ],
            },
          },
          Object {
            "Id": "m1",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "Lambda",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m2",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "Lambda",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "faciaPurgerLambdaServiceRoleD80DA945": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "facia-purger",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/frontend-lambda",
          },
          Object {
            "Key": "Stack",
            "Value": "frontend",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "faciaPurgerLambdaServiceRoleDefaultPolicyD659707F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/frontend/facia-purger",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/frontend/facia-purger/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "faciaPurgerLambdaServiceRoleDefaultPolicyD659707F",
        "Roles": Array [
          Object {
            "Ref": "faciaPurgerLambdaServiceRoleD80DA945",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "invacationAlarmDCBD52AE": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          Object {
            "Ref": "NotificationTopicEB7A0DF1",
          },
        ],
        "AlarmDescription": "Notify if there are less than 5 invocations over last 10 minutes or there is insufficient data (i.e. no invocations)",
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "FunctionName",
            "Value": Object {
              "Ref": "Lambda",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Invocations",
        "Namespace": "AWS/Lambda",
        "Period": 600,
        "Statistic": "Sum",
        "Threshold": 5,
        "TreatMissingData": "breaching",
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;
